File: management/univention-directory-listener/src/filter.c
/*
 * Univention Directory Listener
 *  filter.c
 *
 * Copyright 2004-2019 Univention GmbH
 *
 * https://www.univention.de/
 *
 * All rights reserved.
 *
 * The source code of this program is made available
 * under the terms of the GNU Affero General Public License version 3
 * (GNU AGPL V3) as published by the Free Software Foundation.
 *
 * Binary versions of this program provided by Univention to you as
 * well as other copyrighted, protected or trademarked materials like
 * Logos, graphics, fonts, specific documentations and configurations,
 * cryptographic keys etc. are subject to a license agreement between
 * you and Univention and not subject to the GNU AGPL V3.
 *
 * In the case you use this program under the terms of the GNU AGPL V3,
 * the program is provided in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public
 * License with the Debian GNU/Linux or Univention distribution in file
 * /usr/share/common-licenses/AGPL-3; if not, see
 * <https://www.gnu.org/licenses/>.
 */

/*
 * Functions to match LDAP filters to cache entries. Currently, we don't
 * use any schema information. However, to do this properly, we'd need to.
 */

#define _GNU_SOURCE /* for strndup */

#include <stdio.h>
#include <stdlib.h>
#include <string.h>

#include "common.h"
#include "filter.h"


/* Check if entry matches value of attribute.
 * @param attribute Name of attribute.
 * @param value Expected string value. '*' as a wild-card at the front or end is supported.
 * @param entry Cached LDAP entry to match.
 * @return 1 on match, 0 otherwise.
 */
static int cache_entry_match_attribute_value(char *attribute, char *value, CacheEntry *entry) {
	CacheEntryAttribute **a;
	char **v;
	int len;
	int rv = 0;
	char *substr = NULL;
	int begins = 0, ends = 0;

	for (a = entry->attributes; a != NULL && *a != NULL; a++) {
		if (strcmp((*a)->name, attribute) == 0)
			break;
	}
	if (a == NULL || *a == NULL)
		return 0;

	if (strcmp(value, "*") == 0)
		return 1;

	len = strlen(value);
	begins = value[0] == '*';
	ends = value[len - 1] == '*';
	if (begins || ends)
		substr = strndup(value + begins, len - begins - ends);

	for (v = (*a)->values; v != NULL && *v != NULL; v++) {
		char *match;
		if (strcmp(*v, value) == 0) {
			rv = 1;
			break;
		} else if (substr != NULL && (match = strstr(*v, substr)) != NULL) {
			rv = (begins || match == *v) && (ends || strcmp(match, substr) == 0);
			if (rv)
				break;
		}
	}

	free(substr);
	return rv;
}


/* Check if entry matches the LDAP filter.
 * @param filter LDAP search filter supporting. LIMITED.
 * @param first Index into filter to specify start character.
 * @param last Index into filter to specify last character.
 * @param entry Cached LDAP entry to match.
 * @return 1 on match, 0 on no match, -1 on errors.
 */
static int __cache_entry_ldap_filter_match(char *filter, int first, int last, CacheEntry *entry) {
	/* sanity check */
	if (filter[first] != '(' || filter[last] != ')')
		return -1;

	if (filter[first + 1] == '&' || filter[first + 1] == '|' || filter[first + 1] == '!') {
		int i;
		int begin = -1;
		int depth = 0;

		/* sanity check */
		if (filter[first + 2] != '(' || filter[last - 1] != ')')
			return -1;

		for (i = first + 2; i <= last - 1; i++) {
			if (filter[i] == '(') {
				if (begin == -1)
					begin = i;
				++depth;
			} else if (filter[i] == ')' && begin != -1) {
				int cond_is_true;

				--depth;
				if (depth != 0)
					continue;

				cond_is_true = __cache_entry_ldap_filter_match(filter, begin, i, entry);
				if (!cond_is_true && filter[first + 1] == '&')
					return 0;
				else if (cond_is_true && filter[first + 1] == '|')
					return 1;
				else if (filter[first + 1] == '!')
					return !cond_is_true;

				begin = -1;
			}
		}

		if (filter[first + 1] == '&')
			return 1;
		else if (filter[first + 1] == '|')
			return 0;
		else /* '!', we shouldn't get here */
			return -1;
	} else {
		int type = -1; /* 0: =; 1: ~=; 2: >=; 3: <= */
		char *attr, *val;
		int i;
		int rv;

		for (i = first + 1; i <= last - 1; i++) {
			if (filter[i] == '=' && i > first + 1) {
				if (filter[i - 1] == '~')
					type = 1;
				else if (filter[i - 1] == '>')
					type = 2;
				else if (filter[i - 1] == '<')
					type = 3;
				else
					type = 0;
				break;
			}
		}
		if (type != 0) /* (type == -1) */
			return -1;

		attr = strndup(filter + first + 1, i - first - 1);
		val = strndup(filter + i + 1, last - i - 1);

		rv = cache_entry_match_attribute_value(attr, val, entry);

		free(attr);
		free(val);

		return rv;
	}

	return -1;
}


/* Check if entry matches LDAP dn.
 * @param filter An array of LDAP filters, scopes and bases.
 * @param dn The distinguished name of the cached LDAP entry.
 * @param entry Cached LDAP entry to match.
 * @return 1 on match, 0 otherwise.
 */
int cache_entry_ldap_filter_match(struct filter **filter, const char *dn, CacheEntry *entry) {
	struct filter **f;
	size_t dn_len = strlen(dn);

	for (f = filter; f != NULL && *f != NULL; f++) {
		/* check if base and scope match */
		if ((*f)->base != NULL && (*f)->base[0] != '\0') {
			size_t b_len = strlen((*f)->base);
			/* No match if required base is longer then tested dn */
			if (b_len > dn_len)
				continue;
			/* No match if testes dn does not end on required base */
			if (strcmp(dn + dn_len - b_len, (*f)->base))
				continue;

			switch ((*f)->scope) {
			case LDAP_SCOPE_BASE:
				/* skip if more levels exists. */
				if (strchr(dn, ',') <= dn + dn_len - b_len)
					continue;
				break;
			case LDAP_SCOPE_ONELEVEL:
				/* skip if more then one level */
				if (strchr(dn, ',') + 1 != dn + dn_len - b_len)
					continue;
				break;
			}
		}

		int len = strlen((*f)->filter);
		if (__cache_entry_ldap_filter_match((*f)->filter, 0, len - 1, entry))
			return 1;
	}
	return 0;
}

File: management/univention-directory-manager-modules/modules/univention/admin/filter.py
# -*- coding: utf-8 -*-
# Like what you see? Join us!
# https://www.univention.com/about-us/careers/vacancies/
#
# Copyright 2004-2022 Univention GmbH
#
# https://www.univention.de/
#
# All rights reserved.
#
# The source code of this program is made available
# under the terms of the GNU Affero General Public License version 3
# (GNU AGPL V3) as published by the Free Software Foundation.
#
# Binary versions of this program provided by Univention to you as
# well as other copyrighted, protected or trademarked materials like
# Logos, graphics, fonts, specific documentations and configurations,
# cryptographic keys etc. are subject to a license agreement between
# you and Univention and not subject to the GNU AGPL V3.
#
# In the case you use this program under the terms of the GNU AGPL V3,
# the program is provided in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
# GNU Affero General Public License for more details.
#
# You should have received a copy of the GNU Affero General Public
# License with the Debian GNU/Linux or Univention distribution in file
# /usr/share/common-licenses/AGPL-3; if not, see
# <https://www.gnu.org/licenses/>.

"""
|UDM| functions to parse, modify and create |LDAP| style search filters
"""

import re
from typing import Callable, Iterator, List, Match, Optional, Sequence, TypeVar, Union  # noqa: F401

import six
from ldap.filter import filter_format

import univention.admin.uexceptions

T = TypeVar("T")


class conjunction(object):
	"""
	LDAP filter conjunction (`&`) or disjunction (`|`).
	"""

	OPS = frozenset({'&', '|', '!'})

	def __init__(self, type, expressions):
		# type: (str, List[Union[conjunction, expression]]) -> None
		"""
		Create LDAP filter conjunction or disjunction.

		>>> c = conjunction('&', ['(objectClass=*)'])
		>>> c = conjunction('|', ['(objectClass=*)'])
		"""
		assert type in self.OPS
		self.type = type
		self.expressions = expressions

	@classmethod
	def _parse(cls, text):
		# type: (str) -> conjunction
		op = text[0]
		expressions = [parse(s) for s in cls._split(text[1:])]
		return conjunction(op, expressions)

	@staticmethod
	def _split(text):
		# type: (str) -> Iterator[str]
		depth = 0
		begin = -1
		for i, c in enumerate(text):
			if c == '(':
				depth += 1
				if depth == 1:
					begin = i
			elif c == ')':
				depth -= 1
				if depth == 0 and begin > -1:
					yield text[begin:i + 1]
					begin = -1

	def __str__(self):
		# type: () -> str
		"""
		Return string representation.

		>>> str(conjunction('&', ['(objectClass=*)']))
		'(&(objectClass=*))'
		>>> str(conjunction('|', ['(objectClass=*)']))
		'(|(objectClass=*))'
		>>> str(conjunction('!', ['(objectClass=*)']))
		'(!(objectClass=*))'
		>>> str(conjunction('&', []))
		''
		"""
		if not self.expressions:
			return ''
		return '(%s%s)' % (self.type, ''.join(map(six.text_type, self.expressions)))

	def __unicode__(self):
		# type: () -> str
		return self.__str__()

	def __repr__(self):
		# type: () -> str
		"""
		Return canonical representation.

		>>> conjunction('&', ['(objectClass=*)'])
		conjunction('&', ['(objectClass=*)'])
		>>> conjunction('|', ['(objectClass=*)'])
		conjunction('|', ['(objectClass=*)'])
		"""
		return '%s(%r, %r)' % (self.__class__.__name__, self.type, self.expressions)

	def append_unmapped_filter_string(self, filter_s, rewrite_function, mapping):
		# type: (str, Callable[[expression, Optional[T]], None], T) -> None
		if filter_s:
			filter_p = parse(filter_s)
			walk(filter_p, rewrite_function, arg=mapping)
			self.expressions.append(filter_p)


class expression(object):
	"""
	LDAP filter expression.
	"""

	OPS = frozenset({'=', '>=', '<=', '~=', '=*'} | {'>', '<', '!='})
	# LDAP RFC 4515 + UCS specific extensions
	RE_OP = re.compile(r'([<>]=?|[!~]=|=(?:[*]$)?)')

	def __init__(self, variable='', value='', operator='=', escape=False):
		# type: (str, str, str, bool) -> None
		"""
		Create LDAP filter expression.

		>>> e = expression('objectClass', '*', escape=False)
		>>> e = expression('objectClass', '*', '!=', escape=False)
		>>> e = expression('uidNumber', '10', '<') # < <= > >=
		"""
		assert operator in self.OPS
		if operator == '=' and value == '*':
			operator, value = '=*', ''
		if operator == '=*' and value:
			raise univention.admin.uexceptions.valueInvalidSyntax(value)
		self.variable = variable
		self.value = value
		self.operator = operator
		self._escape = escape

	@classmethod
	def _parse(cls, text):
		# type: (str) -> expression
		var, op, val = cls.RE_OP.split(text, 1)
		return expression(var, val, operator=op)

	def __str__(self):
		# type: () -> str
		r"""
		Return string representation.

		>>> str(expression('objectClass', '*', escape=False))
		'(objectClass=*)'
		>>> str(expression('objectClass', '*', '!=', escape=False))
		'(!(objectClass=*))'
		>>> str(expression('uidNumber', '10', '<'))
		'(!(uidNumber>=10))'
		>>> str(expression('cn', '', '=*'))
		'(cn=*)'
		>>> str(expression('cn', '*', '='))
		'(cn=*)'
		>>> str(expression('cn', r'*\2A*', '='))
		'(cn=*\\2A*)'
		"""
		if self.operator == '<':
			return self.escape('(!(%s>=%s))', (self.variable, self.value))
		elif self.operator == '>':
			return self.escape('(!(%s<=%s))', (self.variable, self.value))
		elif self.operator == '!=':
			return self.escape('(!(%s=%s))', (self.variable, self.value))
		else:
			return self.escape('(%%s%s%%s)' % (self.operator,), (self.variable, self.value))

	def escape(self, string, args):
		# type: (str, Sequence[str]) -> str
		if self._escape:
			return filter_format(string, args)
		return string % args

	def transform_to_conjunction(self, con):
		# type: (conjunction) -> None
		if not isinstance(con, conjunction):
			raise TypeError('must be conjunction, got %s(%r)' % (type(con).__name__, con))
		self.__dict__.clear()
		self.__dict__.update(con.__dict__)
		self.__class__ = type(con)  # type: ignore

	def __unicode__(self):
		# type: () -> str
		return self.__str__()

	def __repr__(self):
		# type: () -> str
		"""
		Return canonical representation.

		>>> expression('objectClass', 'foo*', escape=False)
		expression('objectClass', 'foo*', '=')
		>>> expression('objectClass', '*', '!=', escape=False)
		expression('objectClass', '*', '!=')
		>>> expression('objectClass', '*', '=', escape=False)
		expression('objectClass', '', '=*')
		>>> expression('objectClass', '', '=*', escape=False)
		expression('objectClass', '', '=*')
		"""
		return '%s(%r, %r, %r)' % (self.__class__.__name__, self.variable, self.value, self.operator)


def parse(filter_s, begin=0, end=-1):
	# type: (Union[conjunction, expression, str], int, int) -> Union[conjunction, expression]
	r"""
	Parse LDAP filter string.

	>>> filter_s='(|(&(!(zone=univention.de))(soa=test))(nameserver=bar))'
	>>> parse(filter_s)
	conjunction('|', [conjunction('&', [conjunction('!', [expression('zone', 'univention.de', '=')]), expression('soa', 'test', '=')]), expression('nameserver', 'bar', '=')])
	>>> parse('(!(key>=29))')
	conjunction('!', [expression('key', '29', '>=')])
	>>> parse('(&(key=va\\\\28!\\\\29ue))')
	conjunction('&', [expression('key', 'va\\\\28!\\\\29ue', '=')])
	>>> parse('(cn=Babs Jensen)')
	expression('cn', 'Babs Jensen', '=')
	>>> parse('(!(cn=Tim Howes))')
	conjunction('!', [expression('cn', 'Tim Howes', '=')])
	>>> parse('(&(objectClass=Person)(|(sn=Jensen)(cn=Babs J*)))')
	conjunction('&', [expression('objectClass', 'Person', '='), conjunction('|', [expression('sn', 'Jensen', '='), expression('cn', 'Babs J*', '=')])])
	>>> parse('(o=univ*of*mich*)')
	expression('o', 'univ*of*mich*', '=')
	>>> parse('(seeAlso=)')
	expression('seeAlso', '', '=')
	>>> parse('(cn:caseExactMatch:=Fred Flintstone)')
	expression('cn:caseExactMatch:', 'Fred Flintstone', '=')
	>>> parse('(cn:=Betty Rubble)')
	expression('cn:', 'Betty Rubble', '=')
	>>> parse('(sn:dn:2.4.6.8.10:=Barney Rubble)')
	expression('sn:dn:2.4.6.8.10:', 'Barney Rubble', '=')
	>>> parse('(o:dn:=Ace Industry)')
	expression('o:dn:', 'Ace Industry', '=')
	>>> parse('(:1.2.3:=Wilma Flintstone)')
	expression(':1.2.3:', 'Wilma Flintstone', '=')
	>>> parse('(:DN:2.4.6.8.10:=Dino)')
	expression(':DN:2.4.6.8.10:', 'Dino', '=')
	>>> parse(r'(o=Parens R Us \28for all your parenthetical needs\29)')
	expression('o', 'Parens R Us \\28for all your parenthetical needs\\29', '=')
	>>> parse(r'(cn=*\2A*)')
	expression('cn', '*\\2A*', '=')
	>>> parse(r'(cn=*)')
	expression('cn', '', '=*')
	>>> parse(r'(filename=C:\5cMyFile)')
	expression('filename', 'C:\\5cMyFile', '=')
	>>> parse(r'(bin=\00\00\00\04)')
	expression('bin', '\\00\\00\\00\\04', '=')
	>>> parse(r'(sn=Lu\c4\8di\c4\87)')
	expression('sn', 'Lu\\c4\\8di\\c4\\87', '=')
	>>> parse(r'(1.3.6.1.4.1.1466.0=\04\02\48\69)')
	expression('1.3.6.1.4.1.1466.0', '\\04\\02\\48\\69', '=')
	"""
	# filter is already parsed
	if not isinstance(filter_s, six.string_types):
		return filter_s

	if end == -1:
		end = len(filter_s) - 1

	if filter_s[begin] == '(' and filter_s[end] == ')':
		begin += 1
		end -= 1

	part = filter_s[begin:end + 1]
	try:
		if filter_s[begin] in conjunction.OPS:
			return conjunction._parse(part)
		else:
			return expression._parse(part)
	except (AssertionError, ValueError):
		raise univention.admin.uexceptions.valueInvalidSyntax(part)


def walk(filter_p, expression_walk_function=None, conjunction_walk_function=None, arg=None):
	# type: (Union[conjunction, expression], Optional[Callable[[expression, Optional[T]], None]], Optional[Callable[[conjunction, Optional[T]], None]], Optional[T]) -> None
	"""
	Walk LDAP filter expression tree.

	:param filter_p: expression tree.
	:param expression_walk_function: Callback for expressions.
	:param conjunction_walk_function: Callback for conjunctions.
	:param arg: Argument to the callback functions.

	>>> filter_s = '(|(&(!(zone=univention.de))(soa=test))(nameserver=bar))'
	>>> filter_p = parse(filter_s)
	>>> def trace(e, a): print((a, str(e)))
	>>> walk(filter_p, trace, None, 'e')
	('e', '(zone=univention.de)')
	('e', '(soa=test)')
	('e', '(nameserver=bar)')
	>>> walk(filter_p, None, trace, 'c')
	('c', '(!(zone=univention.de))')
	('c', '(&(!(zone=univention.de))(soa=test))')
	('c', '(|(&(!(zone=univention.de))(soa=test))(nameserver=bar))')
	"""
	if isinstance(filter_p, conjunction):
		for e in filter_p.expressions:
			walk(e, expression_walk_function, conjunction_walk_function, arg)
		if conjunction_walk_function:
			conjunction_walk_function(filter_p, arg)
	elif isinstance(filter_p, expression):
		if expression_walk_function:
			expression_walk_function(filter_p, arg)
	else:
		raise TypeError(type(filter_p))


FQDN_REGEX = re.compile(r'(?:^|\()fqdn=([^)]+)(?:\)|$)')


def replace_fqdn_filter(filter_s):
	# type: (str) -> str
	"""
	Replaces a filter expression for the read-only attribute fqdn. If no
	such expression can be found the unmodified filter is returned.

	>>> replace_fqdn_filter('fqdn=host.domain.tld')
	'(&(cn=host)(associatedDomain=domain.tld))'
	>>> replace_fqdn_filter('(fqdn=host.domain.tld)')
	'(&(cn=host)(associatedDomain=domain.tld))'
	>>> replace_fqdn_filter('fqdn=domain')
	'(|(cn=domain)(associatedDomain=domain))'
	>>> replace_fqdn_filter('(|(fqdn=host.domain.tld)(fqdn=other.domain.tld2))')
	'(|(&(cn=host)(associatedDomain=domain.tld))(&(cn=other)(associatedDomain=domain.tld2)))'
	"""
	if not isinstance(filter_s, six.string_types):
		return filter_s
	return FQDN_REGEX.sub(_replace_fqdn_filter, filter_s)


def _replace_fqdn_filter(match):
	# type: (Match[str]) -> str
	value, = match.groups()
	try:
		host, domain = value.split('.', 1)
		operator = '&'
	except ValueError:
		host = domain = value
		operator = '|'
	return '(%s(cn=%s)(associatedDomain=%s))' % (operator, host, domain)
